dask_jobqueue/cluster_manager.py

```diff
@@ -0,0 +1,63 @@
+import logging
+import threading
+
+from distributed.deploy import Cluster
+
+from distributed.utils import log_errors
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterManager(Cluster):
+    """ Intermediate Cluster object that should lead to a real ClusterManager
+
+    This tries to improve upstream Cluster object and underlines needs for
+    better decoupling between ClusterManager and Scheduler object
+    """
+
+    def __init__(self):
+        self._target_scale = 0
+        self._lock = threading.Lock()
+
+    def _scale(self, n):
+        """ Asynchronously called scale method
+
+        This allows to do every operation with a coherent ocntext
+        """
+        with log_errors(), self._lock:
```

Comment:
> I use a lock in _scale() so this probably means it should not be a corountine, but is the lock use correct?

As written the lock is not necessary, as long as people call `scale` rather than `_scale`.  The `scale` method adds this method onto the event loop:

        self.scheduler.loop.add_callback(self._scale, n)

So any thread can safely call `scale`, which will ask the event loop thread to call `_scale` the next time it is free.  So the `_scale` method will then ever run on the single event loop thread.  There is no need to protect it with a lock.

Generally speaking when using async frameworks like asyncio or tornado it is very rare to use locks.  Instead we protect ourselves by putting all concurrent code on the event loop.

> Should _scale() be made a corountine?

Maybe.  It calls the method `self.scheduler.retire_workers`.  This method is a coroutine that waits for a response from the workers-to-be-closed.  Should `_scale` also wait until it gets this response?  If so then we'll want to `yield/await` that method call:

```python
# self.scheduler.retire_workers(workers=to_close)
yield self.scheduler.retire_workers(workers=to_close)
```

If you do that then you will need to make `_scale` a coroutine as well.  You may not want this though, I'm not sure.  That becomes a design decision.  So I ask a question back to you:

Should `_scale` wait until the workers have closed themselves before calling `scale_down`?

Reply:
> Should _scale wait until the workers have closed themselves before calling scale_down?

Yes, I believe so for clean worker shutdown. Else there is a risk of losing in memory data. From what I understand, currently we ask the scheduler to shutdown worker cleanly, but then just kills them without waiting for it.

Reply:
That's what the code does currently as I read it, yes.

So you would wait on the tornado future returned by `self.scheduler.retire_workers` by calling `yield self.scheduler.retire_workers(...)`.  Once you do that you need to make the `_scale` method a coroutine.

If you want the scale function to block then I recommend using the `distributed.utils.sync` function.  

```python
def scale(...):
    sync(loop, self._scale, ...)
```

This adds the `_scale` coroutine to the event loop, then waits on a `threading.Event` until it has finished.  This must be called from a thread other than the event loop.

Reply:
Thanks, is the code in Cluster.scale()
```python
self.scheduler.loop.add_callback(self.scheduler.retire_workers, workers=to_close)
self.scheduler.loop.add_callback(self.scale_down, to_close)
```
ensuring `scale_down` will be called after `retire_workers` has finished?

Reply:
No.  It will start `retire_workers` and continue that until it hits a `yield` point.  It will start `scale_down` after that (although other things may come in between).

Reply:
> If you want the scale function to block then I recommend using the distributed.utils.sync function

I don't think I want that.

Otherwise, updates done, so all questions answered. Ready to other reviews if Travis build passes.